debian/bongoperf.1
.\" Man page generated from reStructuredText.
.
.TH "BONGOPERF" "1" "January 30, 2015" "3.0" "bongodb-manual"
.SH NAME
bongoperf \- BongoDB Performance Utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBbongoperf\fP is a utility to check disk I/O performance
independently of BongoDB.
.sp
It times tests of random disk I/O and presents the results. You can
use \fBbongoperf\fP for any case apart from BongoDB. The
\fBmmf\fP \fBtrue\fP mode is completely generic. In
that mode it is somewhat analogous to tools such as \fI\%bonnie++\fP (albeit bongoperf is
simpler).
.sp
Specify options to \fBbongoperf\fP using a JavaScript document.
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fI\%bonnie\fP
.IP \(bu 2
\fI\%bonnie++\fP
.IP \(bu 2
\fI\%Output from an example run\fP
.IP \(bu 2
\fI\%Checking Disk Performance with the bongoperf Utility\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH OPTIONS
.INDENT 0.0
.TP
.B bongoperf
.UNINDENT
.INDENT 0.0
.TP
.B bongoperf
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help, \-h
Returns information on the options and use of \fBbongoperf\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B <jsonconfig>
\fBbongoperf\fP accepts configuration options in the form of a
file that holds a \fIJSON\fP document. You must stream the
content of this file into \fBbongoperf\fP, as in the following
operation:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
bongoperf < config
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In this example \fBconfig\fP is the name of a file that holds a JSON
document that resembles the following example:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
{
  nThreads:<n>,
  fileSizeMB:<n>,
  sleepMicros:<n>,
  mmf:<bool>,
  r:<bool>,
  w:<bool>,
  recSizeKB:<n>,
  syncDelay:<n>
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
See the \fI\%Configuration Fields\fP section for documentation of each
of these fields.
.UNINDENT
.SH CONFIGURATION FIELDS
.INDENT 0.0
.TP
.B bongoperf.nThreads
\fIType:\fP Integer.
.sp
\fIDefault:\fP 1
.sp
Defines the number of threads \fBbongoperf\fP will use in the
test. To saturate your system\(aqs storage system you will need
multiple threads. Consider setting \fBnThreads\fP to \fB16\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B bongoperf.fileSizeMB
\fIType:\fP Integer.
.sp
\fIDefault:\fP 1 megabyte (i.e. 1024\s-2\u2\d\s0 bytes)
.sp
Test file size.
.UNINDENT
.INDENT 0.0
.TP
.B bongoperf.sleepMicros
\fIType:\fP Integer.
.sp
\fIDefault:\fP 0
.sp
\fBbongoperf\fP will pause for the number of specified
\fBsleepMicros\fP divided by the
\fBnThreads\fP between each operation.
.UNINDENT
.INDENT 0.0
.TP
.B bongoperf.mmf
\fIType:\fP Boolean.
.sp
\fIDefault:\fP \fBfalse\fP
.sp
Set \fBmmf\fP to \fBtrue\fP to use memory mapped
files for the tests.
.sp
Generally:
.INDENT 7.0
.IP \(bu 2
when \fBmmf\fP is \fBfalse\fP, \fBbongoperf\fP
tests direct, physical, I/O, without caching. Use a large file
size to test heavy random I/O load and to avoid I/O coalescing.
.IP \(bu 2
when \fBmmf\fP is \fBtrue\fP, \fBbongoperf\fP
runs tests of the caching system, and can use normal file system
cache. Use \fBmmf\fP in this mode to test file system cache
behavior with memory mapped files.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B bongoperf.r
\fIType:\fP Boolean.
.sp
\fIDefault:\fP \fBfalse\fP
.sp
Set \fBr\fP to \fBtrue\fP to perform reads as part of
the tests.
.sp
Either \fBr\fP or \fBw\fP must be \fBtrue\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B bongoperf.w
\fIType:\fP Boolean.
.sp
\fIDefault:\fP \fBfalse\fP
.sp
Set \fBw\fP to \fBtrue\fP to perform writes as part of
the tests.
.sp
Either \fBr\fP or \fBw\fP must be \fBtrue\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B bongoperf.recSizeKB
New in version 2.4.

.sp
\fIType:\fP Integer.
.sp
\fIDefault:\fP 4 kb
.sp
The size of each write operation.
.UNINDENT
.INDENT 0.0
.TP
.B bongoperf.syncDelay
\fIType:\fP Integer.
.sp
\fIDefault:\fP 0
.sp
Seconds between disk flushes. \fBbongoperf.syncDelay\fP is
similar to \fI\-\-syncdelay\fP for \fBbongod\fP\&.
.sp
The \fBsyncDelay\fP controls how frequently
\fBbongoperf\fP performs an asynchronous disk flush of the memory
mapped file used for testing. By default, \fBbongod\fP
performs this operation every 60 seconds. Use
\fBsyncDelay\fP to test basic system performance of
this type of operation.
.sp
Only use \fBsyncDelay\fP in conjunction with
\fBmmf\fP set to \fBtrue\fP\&.
.sp
The default value of \fB0\fP disables this.
.UNINDENT
.SH USE
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongoperf < jsonconfigfile
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Replace \fBjsonconfigfile\fP with the path to the \fBbongoperf\fP
configuration. You may also invoke \fBbongoperf\fP in the
following form:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
echo "{nThreads:16,fileSizeMB:10000,r:true,w:true}" | bongoperf
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In this operation:
.INDENT 0.0
.IP \(bu 2
\fBbongoperf\fP tests direct physical random read and write io\(aqs, using
16 concurrent reader threads.
.IP \(bu 2
\fBbongoperf\fP  uses a 10 gigabyte test file.
.UNINDENT
.sp
Consider using \fBiostat\fP, as invoked in the following example to
monitor I/O performance during the test.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
iostat \-xtm 1
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
BongoDB Documentation Project
.SH COPYRIGHT
2011-2015
.\" Generated by docutils manpage writer.
.


NEWS
WiredTiger release 2.9.0, 2016-09-06
------------------------------------

New features and API changes; refer to the API documentation for full details:

* WT-2360	Add nesting and disjunction functionality to cursor joins
* WT-2552	Add a public API to allow custom filesystem implementations
* WT-2738	Remove the ability to change the default checkpoint name. See the upgrading documentation for details.
* WT-2711	Change the statistics log configuration options. Restrict where statistics log files can be created to remove security vulnerability. See upgrading documentation for details.
* WT-2764	Enhance the checkpoint implementation, which led to a change in default eviction settings. See the upgrading documentation for details.
* WT-2880	Add Zstandard compression support
* See the upgrading documentation for information about renamed and removed statistics.

Significant changes and bug fixes:

* WT-2026	Fix a bug where in-memory pages were allowed to grow too large
* WT-2343	Ensure we cannot rename/drop when a backup cursor is open
* WT-2520	Ensure WT_SESSION::verify does not alter tables
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2664	Change eviction so any eviction thread can find candidates
* WT-2702	Resolve an issue where under high thread load, WiredTiger exceeds cache size
* WT-2731	Resolve an issue where raw compression can create pages that are larger than expected
* WT-2737	Allow reconciliation to scrub dirty pages rather than evicting them
* WT-2757	Fix a bug in column stores where it was not possible to retrieve the new record number after an append
* WT-2816	Improve eviction performance

Issues fixed in MongoDB:

* SERVER-23504	Resolve a resource leak
* SERVER-23588	Stop using _open_osfhandle on Windows
* SERVER-23659	Improve log error message at startup
* SERVER-23661	Fix an issue where $sample takes disproportionately long time on newly created collection
* SERVER-24306	Fix stall in log_flush switching to new files
* SERVER-24580	Add more eviction stats to track efficiency
* SERVER-24580	Enhance eviction when application threads are contributing
* SERVER-24971	Don't cache buffers after application eviction
* SERVER-25843	Remove a redundant test
* SERVER-25845	Resolve an explicit null dereferenced
* SERVER-25846	Resolve a dereference after null with a check
* SERVER-26753	Don't spin on a read-lock in a tight loop

Other noteworthy changes since the previous release:

* WT-2103	Add incremental backup testing to test/format
* WT-2330	Ensure In-memory configurations do not create on-disk collection files
* WT-2450	Resolve an issue where salvage releases pages, then explicitly evicts them. Which can race with the eviction server
* WT-2453	Add multiple eviction queues
* WT-2504	Remove READONLY conditional for base config
* WT-2505	Resolve clang analyzer warnings
* WT-2508	Ensure test programs should remove test directories on the "clean" target
* WT-2512	Change how wtperf throttle values per thread are calculated
* WT-2513	Fix a type conversion warning
* WT-2517	Resolve incorrect uses of setvbuf on Windows
* WT-2518	Optimization of LSM checkpoint handle acquisition
* WT-2522	Resolve incorrect format code in message
* WT-2525	Miscellaneous cleanups of In-memory configurations
* WT-2526	Prevent mixing and matching readonly file-handles and read/write data-handles
* WT-2528	Fix a style error in WiredTiger build
* WT-2529	Be less aggressive asserting in readonly connections
* WT-2531	Resolve a case where in-memory tables are wasting space in truncation
* WT-2532	Change the WT_STREAM_APPEND and WT_STREAM_LINE_BUFFER flags as they overlap
* WT-2533	Don't let in-memory tables return a zero size
* WT-2534	Use atomic add when allocating transaction IDs
* WT-2535	Extend test/format to test for transactions reading their writes
* WT-2537	Check for an old and new uninitialized LSN for recovery
* WT-2539	Make streams a separate handle type from files
* WT-2540	Separate stream and file handle methods
* WT-2542	Resolve fixed-length column store reconciliation overwriting original values
* WT-2544	Fix eviction stats when clear is used
* WT-2546	Change evict server to be smarter when it decides whether or not to evict pages
* WT-2547	Add eviction testing configurations for 1 thread
* WT-2548	Cap the amount of data handed to raw compression
* WT-2550	Add support for raw mode cursors in java
* WT-2553	Add some In-memory documentation and stress testing
* WT-2556	Fix a typo in the Java example code
* WT-2557	Ensure that we truncate logs before closing backup cursor
* WT-2558	Reorganize WT_PAGE structure
* WT-2559	Resolve a potential segfault in Windows logging code
* WT-2560	Add a rwlock to protect transaction state, don't spin on this lock
* WT-2562	Add sleep loop to adapt for slow machines like PPC
* WT-2565	Further limit the amount of data passed to raw compression
* WT-2566	Add explicit memory barriers to Lock/unlock operations
* WT-2567	Add a check to see if logging is enabled for truncate rather than assuming it is
* WT-2568	Add a backward compatible constructor for PackInputStream
* WT-2569	Ensure __win_handle_read is followed by a call to GetLastError()
* WT-2570	Minor lint and cleanups to code
* WT-2571	Join code clean up
* WT-2572	Ensure test/format doesn't select an incompatible set of options for in-memory options configured
* WT-2573	Resolve a potential case where a double free can occur
* WT-2574	Ensure test/format frees all allocated configure memory
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2577	Resolve a potential double-free following an in-memory split
* WT-2579	Change so diagnostic only code can use file streams directly
* WT-2580	Fix potential SWIG naming conflict in Java
* WT-2581	Resolve a case where we fail to track saved updates during a page split
* WT-2582	Fix a case where eviction can return WT_RESTART when running in-memory
* WT-2584	Change to not use periods at the end of error messages

src/bongo/client/examples/bongoperf.cpp
/*    Copyright 2009 10gen Inc.
 *
 *    This program is free software: you can redistribute it and/or  modify
 *    it under the terms of the GNU Affero General Public License, version 3,
 *    as published by the Free Software Foundation.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU Affero General Public License for more details.
 *
 *    You should have received a copy of the GNU Affero General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *    As a special exception, the copyright holders give permission to link the
 *    code of portions of this program with the OpenSSL library under certain
 *    conditions as described in each individual source file and distribute
 *    linked combinations including the program with the OpenSSL library. You
 *    must comply with the GNU Affero General Public License in all respects
 *    for all of the code used other than as permitted herein. If you modify
 *    file(s) with this exception, you may extend this exception to your
 *    version of the file(s), but you are not obligated to do so. If you do not
 *    wish to do so, delete this exception statement from your version. If you
 *    delete this exception statement from all source files in the program,
 *    then also delete it in the license file.
 */

/*
   How to build and run:

   scons bongoperf
   ./bongoperf -h
*/


// note: bongoperf is an internal bongodb utility
// so we define the following macro
#define BONGO_EXPOSE_MACROS 1

#include "bongo/platform/basic.h"

#include <iostream>

#include <boost/filesystem/operations.hpp>

#include "bongo/db/jsobj.h"
#include "bongo/db/json.h"
#include "bongo/db/storage/mmap_v1/logfile.h"
#include "bongo/db/storage/mmap_v1/mmap.h"
#include "bongo/platform/atomic_word.h"
#include "bongo/stdx/thread.h"
#include "bongo/util/allocator.h"
#include "bongo/util/bongoutils/str.h"
#include "bongo/util/processinfo.h"
#include "bongo/util/scopeguard.h"
#include "bongo/util/time_support.h"
#include "bongo/util/timer.h"

using namespace std;
using namespace bongo;
using namespace bongoutils::str;

int dummy;
unsigned recSizeKB;
LogFile* lf = 0;
MemoryMappedFile* mmfFile;
char* mmf = 0;
bo options;
unsigned long long len;  // file len
const unsigned PG = 4096;
unsigned nThreadsRunning = 0;

// as this is incremented A LOT, at some point this becomes a bottleneck if very high ops/second (in
// cache) things are happening.
AtomicUInt32 iops;

SimpleMutex m;

int syncDelaySecs = 0;

void syncThread() {
    while (1) {
        bongo::Timer t;
        mmfFile->flush(true);
        cout << "                                                     mmf sync took " << t.millis()
             << "ms" << endl;
        sleepsecs(syncDelaySecs);
    }
}

void stripTrailing(std::string& s, const char* chars) {
    std::string::iterator to = s.begin();
    for (std::string::iterator i = s.begin(); i != s.end(); i++) {
        // During each iteration if i finds a non-"chars" character it writes it to the
        // position of t. So the part of the string left from the "to" iterator is already
        // "cleared" string.
        if (!contains(chars, *i)) {
            if (i != to)
                s.replace(to, to + 1, 1, *i);
            to++;
        }
    }
    s.erase(to, s.end());
}

char* round(char* x) {
    size_t f = (size_t)x;
    char* p = (char*)((f + PG - 1) / PG * PG);
    return p;
}

struct Aligned {
    char x[8192];
    char* addr() {
        return round(x);
    }
};

unsigned long long rrand() {
    // RAND_MAX is very small on windows
    return (static_cast<unsigned long long>(rand()) << 15) ^ rand();
}

void workerThread() {
    bool r = options["r"].trueValue();
    bool w = options["w"].trueValue();
    cout << "read:" << r << " write:" << w << endl;
    long long su = options["sleepMicros"].numberLong();
    Aligned a;
    while (1) {
        unsigned long long rofs = (rrand() * PG) % len;
        unsigned long long wofs = (rrand() * PG) % len;
        const unsigned P = PG / 1024;
        if (mmf) {
            if (r) {
                for (unsigned p = P; p <= recSizeKB; p += P) {
                    if (rofs < len)
                        dummy += mmf[rofs];
                    rofs += PG;
                }
                iops.fetchAndAdd(1);
            }
            if (w) {
                for (unsigned p = P; p <= recSizeKB; p += P) {
                    if (wofs < len)
                        mmf[wofs] = 3;
                    wofs += PG;
                }
                iops.fetchAndAdd(1);
            }
        } else {
            if (r) {
                lf->readAt(rofs, a.addr(), recSizeKB * 1024);
                iops.fetchAndAdd(1);
            }
            if (w) {
                lf->writeAt(wofs, a.addr(), recSizeKB * 1024);
                iops.fetchAndAdd(1);
            }
        }
        long long micros = su / nThreadsRunning;
        if (micros) {
            sleepmicros(micros);
        }
    }
}

void go() {
    verify(options["r"].trueValue() || options["w"].trueValue());

    recSizeKB = options["recSizeKB"].numberInt();
    if (recSizeKB == 0)
        recSizeKB = 4;
    verify(recSizeKB <= 64000 && recSizeKB > 0);

    auto txn = cc().makeOperationContext();
    MemoryMappedFile f(txn.get());
    ON_BLOCK_EXIT([&f, &txn] {
        LockBongoFilesExclusive lock(txn.get());
        f.close(txn.get());
    });

    cout << "creating test file size:";
    len = options["fileSizeMB"].numberLong();
    if (len == 0)
        len = 1;
    cout << len << "MB ..." << endl;

    if (0 && len > 2000 && !options["mmf"].trueValue()) {
        // todo make tests use 64 bit offsets in their i/o -- i.e. adjust LogFile::writeAt and such
        cout << "\nsizes > 2GB not yet supported with mmf:false" << endl;
        return;
    }
    len *= 1024 * 1024;
    const char* fname = "./bongoperf__testfile__tmp";
    try {
        boost::filesystem::remove(fname);
    } catch (...) {
        cout << "error deleting file " << fname << endl;
        return;
    }
    lf = new LogFile(fname, true);
    // needs to be big as we are using synchronousAppend.  if we used a regular BongoFile it
    // wouldn't have to be
    const unsigned sz = 1024 * 1024 * 32;
    char* buf = (char*)bongoMalloc(sz + 4096);
    const char* p = round(buf);
    for (unsigned long long i = 0; i < len; i += sz) {
        lf->synchronousAppend(p, sz);
        if (i % (1024ULL * 1024 * 1024) == 0 && i) {
            cout << i / (1024ULL * 1024 * 1024) << "GB..." << endl;
        }
    }
    BSONObj& o = options;

    if (o["mmf"].trueValue()) {
        delete lf;
        lf = 0;
        mmfFile = new MemoryMappedFile(txn.get());
        mmf = (char*)mmfFile->map(txn.get(), fname);
        verify(mmf);

        syncDelaySecs = options["syncDelay"].numberInt();
        if (syncDelaySecs) {
            stdx::thread t(syncThread);
            t.detach();
        }
    }

    cout << "testing..." << endl;

    cout << "options:" << o.toString() << endl;
    unsigned wthr = 1;
    if (!o["nThreads"].eoo()) {
        wthr = (unsigned)o["nThreads"].Int();
    }
    cout << "wthr " << wthr << endl;

    if (wthr < 1) {
        cout << "bad threads field value" << endl;
        return;
    }
    unsigned i = 0;
    unsigned d = 1;
    unsigned& nthr = nThreadsRunning;
    while (1) {
        if (i++ % 8 == 0) {
            if (nthr < wthr) {
                while (nthr < wthr && nthr < d) {
                    nthr++;
                    stdx::thread w(workerThread);
                    w.detach();
                }
                cout << "new thread, total running : " << nthr << endl;
                d *= 2;
            }
        }
        sleepsecs(1);
        unsigned long long w = iops.loadRelaxed();
        iops.store(0);
        w /= 1;  // 1 secs
        cout << w << " ops/sec ";
        if (mmf == 0)
            // only writing 4 bytes with mmf so we don't say this
            cout << (w * PG / 1024 / 1024) << " MB/sec";
        cout << endl;
    }
}

int main(int argc, char* argv[]) {
    try {
        cout << "bongoperf" << endl;

        if (argc > 1) {
            cout <<

                "\n"
                "usage:\n"
                "\n"
                "  bongoperf < myjsonconfigfile\n"
                "\n"
                "  {\n"
                "    nThreads:<n>,     // number of threads (default 1)\n"
                "    fileSizeMB:<n>,   // test file size (default 1MB)\n"
                "    sleepMicros:<n>,  // pause for sleepMicros/nThreads between each operation "
                "(default 0)\n"
                "    mmf:<bool>,       // if true do i/o's via memory mapped files (default "
                "false)\n"
                "    r:<bool>,         // do reads (default false)\n"
                "    w:<bool>,         // do writes (default false)\n"
                "    recSizeKB:<n>,    // size of each write (default 4KB)\n"
                "    syncDelay:<n>     // secs between fsyncs, like --syncdelay in bongod. "
                "(default 0/never)\n"
                "  }\n"
                "\n"
                "bongoperf is a performance testing tool. the initial tests are of disk subsystem "
                "performance; \n"
                "  tests of bongos and bongod will be added later.\n"
                "most fields are optional.\n"
                "non-mmf io is direct io (no caching). use a large file size to test making the "
                "heads\n"
                "  move significantly and to avoid i/o coalescing\n"
                "mmf io uses caching (the file system cache).\n"
                "\n"

                 << endl;
            return EXIT_SUCCESS;
        }

        cout << "use -h for help" << endl;

        char input[1024];
        memset(input, 0, sizeof(input));
        cin.read(input, 1000);
        if (*input == 0) {
            cout << "error no options found on stdin for bongoperf" << endl;
            return EXIT_FAILURE;
        }

        string s = input;
        stripTrailing(s, " \n\r\0x1a");
        try {
            options = fromjson(s);
        } catch (...) {
            cout << "couldn't parse json options. input was:\n|" << s << "|" << endl;
            return EXIT_FAILURE;
        }
        cout << "parsed options:\n" << options.toString() << endl;
        ProcessInfo::initializeSystemInfo();

        go();
    } catch (DBException& e) {
        cout << "caught DBException " << e.toString() << endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
